### PR TITLE
fix(gateway): Resolve gateway retry issues (`UplinkFetcher failed to update...`) (`version-0.x`)

### PR DESCRIPTION
**Pull request by trevor-scheer**
Currently the gateway implements incorrect retry logic. The `null` response case (no schema change found) is handled as a retry-able result (or failure case) when really it should be handled as a success and result in completing the current retry sequence with no errors thrown.

It's worth mentioning that this issue should have never affected gateway load/startup (only update), since the gateway has no `ifAfterId` to send to our server and thus should never expect an `Unchanged` result from its initial request. It should also have never affected gateway _updates_ since in the update case the result would be non-null and escape the retry case. All this is to say, gateway function should have been unaffected due to this issue, but it did log errors even from the happy path (successful retry after an error).

Additionally, this PR introduces the `async-retry` package which wraps the retry logic for us and handles exponential backoff.

### CODE REVIEW

**Review by glasser — APPROVED** [review]

File: gateway-js/src/supergraphManagers/UplinkFetcher/__tests__/loadSupergraphSdlFromStorage.test.ts [hunk 1]
```diff
@@ -368,3 +370,28 @@ describe('loadSupergraphSdlFromStorage', () => {
   });
 });
 
+
+describe("loadSupergraphSdlFromUplinks", () => {
+  beforeEach(nockBeforeEach);
+  afterEach(nockAfterEach);
+
+  it("doesn't retry in the unchanged / null case", async () => {
```
[the remark this answers is not in the document]
Comment (trevor-scheer) [reply gateway-js/src/supergraphManagers/UplinkFetcher/__tests__/loadSupergraphSdlFromStorage.test.ts:378]: Yep (confirmable via `git checkout 8de3c6b && npm t`)

File: gateway-js/src/supergraphManagers/UplinkFetcher/loadSupergraphSdlFromStorage.ts [hunk 2]
```diff
@@ -56,33 +55,45 @@ export async function loadSupergraphSdlFromUplinks({
...
-  }
-  if (result === null && lastException !== null) {
+
+      // If we make it here, we didn't throw on this attempt. We'll clear
+      // any error we might've captured so we don't throw it below.
+      lastException = null;
+      return result;
+    },
+    {
+      retries: maxRetries,
```
[the remark this answers is not in the document]
Comment (trevor-scheer) [reply gateway-js/src/supergraphManagers/UplinkFetcher/loadSupergraphSdlFromStorage.ts:83]: 39b9ffb

File: gateway-js/src/supergraphManagers/UplinkFetcher/loadSupergraphSdlFromStorage.ts [hunk 3]
```diff
@@ -56,33 +55,45 @@ export async function loadSupergraphSdlFromUplinks({
...
-        compositionId
+        compositionId,
       });
-    } catch (e) {
-      lastException = e;
-    }
-  }
-  if (result === null && lastException !== null) {
+
+      // If we make it here, we didn't throw on this attempt. We'll clear
```
[the remark this answers is not in the document]
Comment (trevor-scheer) [reply gateway-js/src/supergraphManagers/UplinkFetcher/loadSupergraphSdlFromStorage.ts]: 39b9ffb

File: gateway-js/src/supergraphManagers/UplinkFetcher/index.ts [hunk 4]
```diff
@@ -81,6 +82,7 @@ export class UplinkFetcher implements SupergraphManager {
       fetcher: this.config.fetcher,
       compositionId: this.compositionId ?? null,
       maxRetries: this.config.maxRetries,
+      roundRobinSeed: this.fetchCount,
```
[the remark this answers is not in the document]
Comment (trevor-scheer) [reply gateway-js/src/supergraphManagers/UplinkFetcher/index.ts]: 39b9ffb

**Pull request merged**